### PR TITLE
Insert trampoline instruction for EBPF instrumentation framework 

### PR DESCRIPTION
**Pull request by usamasaqib**
### What does this PR do?
This PR is a first in a series of PRs to implement an EBPF instrumentation system. The purpose of this system will be to allow small code blocks to be run before the execution of all eBPF programs in a binary.
   
This PR:
1. Adds a flag to pass `-pg` option to clang when compiling `usm` `tracer` and `conntrack` binaries. More details below.
2. Changes the `Compile` method of the runtime compiler to accept flags for llc.
3. Adds tests to verify that the expected changes in the bytecode are present when the new options are used.

This PR provides the supporting infrastructure for inserting a trampoline instruction which will be used to jump to the instrumentation code. Subsequent PRs will demonstrate the usecases for this system
The trampoline instruction is compiler generated by passing the `-pg` option. This option instruments a `call -1` instruction in the beginning of the eBPF bytecode.
Subsequent PRs will patch this instruction with an absolute jump to the end of the bytecode, where the instrumentation code will reside.

When instrumentation is enabled we will add a new section to the eBPF ELF files named `.build.instrumentation` with the contents set to the null-terminated string `enabled`. This will easily allow us to inspect the ELF file and determine whether it is expecting to be patched with instrumentation code.

The PR makes changes to the runtime compilation code to allow a module to pass flags to llc. This facility will be used in later PRs making use of this framework.

The PR also provides tests to verify that all eBPF programs are correctly instrumented with the trampoline patch point, when instrumentation is enabled.

### Motivation
The primary motivation for this PR is to optimize the eBPF telemetry functionality. The system introduced here will be used to limit all telemetry map operation to one lookup per program invocation

However, the system is meant to be generic and can be used to perform different kinds of "global" operations before eBPF programs. These can include:
- Pause/resume all eBPF programs.
- Measuring eBPF program statistics on kernels without this facility.
- And most likely many more...

### Describe how to test/QA your changes
Relevant unit tests are added.

### CODE REVIEW

**Review by hmahmood — APPROVED** [review]

**Review by L3n41c — APPROVED** [review]
LGTM for files owned by @DataDog/container-integrations

File: pkg/network/tracer/compile.go [hunk 1]
```diff
@@ -32,5 +32,12 @@ func getCFlags(config *config.Config) []string {
 	if config.BPFDebug {
 		cflags = append(cflags, "-DDEBUG=1")
 	}
+	if config.EBPFInstrumentationEnabled {
```
Comment (val06) [inline pkg/network/tracer/compile.go:35]: please add a comment (you can copy paste from PR description) to explain the purpose of this flag

File: tasks/system_probe.py [hunk 2]
```diff
@@ -96,7 +98,7 @@ def ninja_define_ebpf_compiler(nw, strip_object_files=False, kernel_release=None
     strip = "&& llvm-strip -g $out" if strip_object_files else ""
     nw.rule(
         name="llc",
-        command=f"llc -march=bpf -filetype=obj -o $out $in {strip}",
+        command=f"llc -stack-size-section -march=bpf -filetype=obj -o $out $in {strip}",
```
Comment (val06) [inline tasks/system_probe.py:101]: why do we need this? and if this flag is helpful, maybe we should add it as a standalone PR (not related to the ebpf instrumentation framework)

File: pkg/ebpf/telemetry/errors_telemetry.go [hunk 3]
```diff
@@ -256,3 +259,31 @@ func ebpfTelemetrySupported() (bool, error) {
...
+	objFile, err := elf.NewFile(bytecode)
+	if err != nil {
+		return false, fmt.Errorf("failed to open elf file: %w", err)
+	}
+	defer objFile.Close()
+
+	const instrumentationSectionName = ".build.instrumentation"
+	sec := objFile.Section(instrumentationSectionName)
+	if sec == nil {
+		return false, nil
```
Comment (val06) [inline pkg/ebpf/telemetry/errors_telemetry.go:273]: if that's not an error, please state explicitly in the comment

**Review by val06 — CHANGES_REQUESTED** [review]
why are we using `call -1` as a marker?
that way even if we don't want to have any logic we need to add a trampoline and then retpoline to resume the regular execution.

can we comeup with a placeholder that is equivalent to NOPs?
Windows kernel [example](https://www.codeproject.com/Articles/737907/Taking-advantage-of-Windows-Hot-Patching-mechanism)

**Comment by usamasaqib** [comment]
> why are we using `call -1` as a marker? that way even if we don't want to have any logic we need to add a trampoline and then retpoline to resume the regular execution.
> 
> can we comeup with a placeholder that is equivalent to NOPs? Windows kernel [example](https://www.codeproject.com/Articles/737907/Taking-advantage-of-Windows-Hot-Patching-mechanism)

We do not have control over the marker. The compiler instruments a `call -1` when `-pg` is specified. We use this as a patch point to add our trampoline instruction.

The retpoline will not be used if no instrumentation is enabled. That is just an example of the most basic instrumentation.
If no instrumentation is enabled we will patch the `call -1` with a nop instruction.

File: tasks/system_probe.py (hunk 2 above)
Comment (usamasaqib) [reply tasks/system_probe.py:101]: This flag is required for ebpf telemetry support. Will add it with that PR.

File: pkg/network/tracer/compile.go (hunk 1 above)
Comment (usamasaqib) [reply pkg/network/tracer/compile.go:35]: I have described the config option where it is defined
https://github.com/DataDog/datadog-agent/pull/23070/files#diff-7a70d1343f42219c3b2c07d3bf90635f3d79b9e4fa7ed18ede433218e25e929eR82-R83

File: pkg/ebpf/telemetry/errors_telemetry.go (hunk 3 above)
Comment (usamasaqib) [reply pkg/ebpf/telemetry/errors_telemetry.go:275]: Done!

File: pkg/network/usm/compile.go [hunk 4]
```diff
@@ -18,7 +18,7 @@ import (
 //go:generate $GOPATH/bin/integrity pkg/ebpf/bytecode/build/runtime/usm.c pkg/ebpf/bytecode/runtime/usm.go runtime
 
 func getRuntimeCompiledUSM(config *config.Config) (runtime.CompiledOutput, error) {
-	return runtime.Usm.Compile(&config.Config, getCFlags(config), statsd.Client)
+	return runtime.Usm.Compile(&config.Config, getCFlags(config), getLlcFlags(), statsd.Client)
```
Comment (guyarb) [inline pkg/network/usm/compile.go:21]: why not simply passing nil?
I saw you did it from oom_kill and tcp_queue_length

File: pkg/network/usm/sharedlibraries/compile.go [hunk 5]
```diff
@@ -18,7 +18,7 @@ import (
 //go:generate $GOPATH/bin/integrity pkg/ebpf/bytecode/build/runtime/shared-libraries.c pkg/ebpf/bytecode/runtime/shared-libraries.go runtime
 
 func getRuntimeCompiledSharedLibraries(config *config.Config) (runtime.CompiledOutput, error) {
-	return runtime.SharedLibraries.Compile(&config.Config, getCFlags(config), statsd.Client)
+	return runtime.SharedLibraries.Compile(&config.Config, getCFlags(config), getLlcFlags(), statsd.Client)
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/compile.go:21]: same comment as above

File: pkg/network/usm/compile_test.go [hunk 6]
```diff
@@ -32,3 +35,41 @@ func TestHttpCompile(t *testing.T) {
 		_ = out.Close()
 	})
 }
+
+func TestUSMCorrectlyInstrumentedWithTrampoline(t *testing.T) {
+	ebpftest.TestBuildMode(t, ebpftest.RuntimeCompiled, "", func(t *testing.T) {
```
Comment (guyarb) [inline pkg/network/usm/compile_test.go:40]: is the feature only supported in runtime compilation? do we plan to add it to co-re?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/usm/compile.go (hunk 4 above)
Comment (usamasaqib) [reply pkg/network/usm/compile.go:21]: Follow up PRs will introduce a flag to be passed to `llc`

File: pkg/network/usm/compile_test.go (hunk 6 above)
Comment (usamasaqib) [reply pkg/network/usm/compile_test.go:40]: It is supported for all buildmodes.
Prebuilt and CO-RE are tested here: https://github.com/DataDog/datadog-agent/pull/23070/files#diff-e577568bbd8afdcdbd633c3c612d9d60e2a8a84e8370fbf08e5b51d598492dc1R58

Runtime is tested separately here to easily get access to the runtime compilation API.

**Review by val06 — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

File: pkg/network/usm/sharedlibraries/compile.go (hunk 5 above)
Comment (usamasaqib) [reply pkg/network/usm/sharedlibraries/compile.go:21]: Same as above, this function will return valid `llc` flags in subsequent PRs.

**Review by guyarb — APPROVED** [review]

**Pull request merged**